src/potentials.h

```diff
@@ -8,27 +8,138 @@
+ *
+ * The very same format is used as for a homogeneous interaction specified directly on an atom type.
+ */
+struct InteractionData {
+    std::array<AtomData::Tid, 2> atom_id;
+    AtomData interaction;
+};
+
+void from_json(const json &j, std::vector<InteractionData> &interactions);
+void to_json(json &j, const std::vector<InteractionData> &interactions);
+
+/**
+ * @brief Known named combination rules for parameters of pair potential interaction.
+ *
+ * When adding a new one, add a json mapping. Also consider appending the PairMixer::getCombinator()
+ * method to recognize the new rule.
+ */
+enum ECombinationRule {COMB_UNDEFINED = 0, COMB_ARITHMETIC, COMB_GEOMETRIC, COMB_LORENTZ_BERTHELOT};
+NLOHMANN_JSON_SERIALIZE_ENUM( ECombinationRule, {
+    {COMB_UNDEFINED, "undefined"},
+    {COMB_ARITHMETIC, "arithmetic"},
+    {COMB_GEOMETRIC, "geometric"},
+    {COMB_LORENTZ_BERTHELOT, "lorentz_berthelot"},
+    {COMB_LORENTZ_BERTHELOT, "LB"}, // alternative non-canonical name
+})
+
+/**
+ * @brief Exception for handling pair potential initialization.
+ */
+struct PairPotentialException : public std::runtime_error {
+    PairPotentialException(const std::string msg) : std::runtime_error (msg) {};
+};
+
+/**
+ * @brief PairMixer creates a matrix of pair potential coefficients based on the atom properties
+ * and/or custom values using an arbitrary combination rule.
+ *
+ * PairMixer holds three functions that are applied in order extractor → combinator → modifier to create
+ * a coefficient matrix for all possible interactions. The function createPairMatrix applies the functions
+ * on all atom type pairs, and optionally also on the list of custom pair parameters (not the combinator
+ * function).
+ */
+class PairMixer {
+    TExtractorFunc extractor;   //!< Function extracting the coefficient from the AtomData structure
+    TCombinatorFunc combinator; //!< Function combining two values
+    TModifierFunc modifier;     //!< Function modifying the result for fast computations, e.g., a square of
+
+  public:
+    PairMixer(TExtractorFunc extractor, TCombinatorFunc combinator, TModifierFunc modifier = &modIdentity) :
+            extractor(extractor), combinator(combinator), modifier(modifier) {};
+
+    //! @return a square matrix of atoms.size()
+    TPairMatrixPtr createPairMatrix(const std::vector<AtomData> &atoms);
+    //! @return a square matrix of atoms.size()
+    TPairMatrixPtr createPairMatrix(const std::vector<AtomData> &atoms,
+                                    const std::vector<InteractionData> &interactions);
+
+    enum ECoefficient {COEF_ANY, COEF_SIGMA, COEF_EPSILON};
+    static TCombinatorFunc getCombinator(ECombinationRule combination_rule, ECoefficient = COEF_ANY);
+
+    // when explicit custom pairs are the only option
+    inline static double combUndefined(double, double) { return std::numeric_limits<double>::signaling_NaN(); };
+    inline static double combSum(double a, double b) { return (a+b); }
```

Comment:
use `std::plus`

Reply:
I removed the function completely as it is not used at all.